kubeadm release notes draft for 1.6

lukemarsden opened the pull request:
For discussion in SIG-cluster-lifecycle.

Hailing @jbeda @luxas @pires @errordeveloper

caseydavenport commented on release-1.6/release-notes-draft.md:
```diff
@@ -58,12 +58,33 @@
 ### Breaking Changes
+- kubeadm creates RBAC-enabled clusters. This means that some add-ons which
+  worked previously won't work without having their YAML configuration updated.
+  For example, to use Weave Net with a Kubernetes 1.6 cluster created by
```
Why are we calling out Weave Net specifically in the release notes for kubeadm? Please stop doing this. No network provider has special status.

This should be worded like so:

```
kubeadm creates RBAC-enabled clusters. This means that some add-ons which
worked previously won't work without having their YAML configuration updated.
```

Or if you really want to add an example:

```
kubeadm creates RBAC-enabled clusters. This means that some add-ons which
worked previously won't work without having their YAML configuration updated.
For example, CNI network providers will require updated manifests to include RBAC policies.
```

liggitt replied:
> This means that some add-ons which worked previously won't work without having their YAML configuration updated.

The add-ons can either update to include role definitions and bindings (something we'll be hitting hard in 1.7), or the cluster admin can grant permissions to the addon's service account (most use the `default` service account in the `kube-system` namespace currently).

lukemarsden replied:
@caseydavenport I'm trying to provide _useful examples for users_. Many users I've seen have downloaded yamls for various add-ons and will naturally want to reuse them for upgraded clusters. If we don't provide some information on where to find the 1.6-compatible yamls for network providers, users are going to be stuck. Now, arguably the release notes are the wrong place to do this, however I think we should at least have a section "1.6 compatible network add-ons" on the add-ons page and link to it from here and from the kubeadm guide. WDYT?

caseydavenport replied:
> Now, arguably the release notes are the right place to alert people to this (it's a breaking changes section) but the wrong place to list all available 1.6-compatible add-ons

Yep.  Totally agree - we can alert people here, but listing _all_ add-ons here is wrong, as is calling out just one or two specifically.

> 1.6 compatible network add-ons

I think the wording here and in the kubeadm guide should be clear about what has changed in v1.6 and the impact that might have on add-ons in general, then direct users to look at the documentation for their chosen add-on - this is a more scalable solution than maintaining some sort of matrix for picking the right yaml file.

jbeda commented:
Some comments:
* We should call out as a new feature that users can manage tokens (including setting expiration times) with kubeadm.  This and RBAC are, I think, the top line new features.
* The command line flags have been changed.  It might be worthwhile to document what has been removed and added.
* The cloud provider flag has been removed and moved to the config file.  Reasoning is that it is difficult to set up properly and making it a command line flag on kubeadm makes it look easier than it is.
* We are now using new labels for tainting and labeling the master.  @luxas -- do you have the old/new values off the top of your head?  If not, I can look them up.
* On the network/CNI stuff, both RBAC and the new taints/labels will require changes for most CNI configuration files.

luxas commented on release-1.6/release-notes-draft.md:
```diff
@@ -58,12 +58,33 @@
+  server also.
+- kubeadm now supports managing the bootstrap tokens active in your cluster
+  using the new [`kubeadm token`](https://kubernetes.io/docs/admin/bootstrap-tokens/) subcommands.
+- Many other bug-fixes and improvements have been contributed that make using
```
* The kubelet API is now secured, only cluster admins are allowed to access it.
* Insecure access to the API Server over `localhost:8080` is disabled by default
* The control plane components now talk securely to each other. The API Server talks securely to the kubelets in the cluster
* A new label and taint is used for marking the master. The label is `node-role.kubernetes.io/master=""` and the taint has the effect `NoSchedule`
* Added experimental self-hosting support
* Supporting only etcd3, with 3.0.14 as the minimum version
* Added an experimental command for running only one of kubeadm's functions, for instance `kubeadm alpha phase certs selfsign` generates all necessary certificates for a cluster. You can use these building block functions in your own, custom deployment. In future versions, this command will mature and be moved to beta quality.
* Debian packages are built for ppc64le and s390x as well

Some of these are breaking as well:
* A new label and taint is used for marking the master. The label is `node-role.kubernetes.io/master=""` and the taint has the effect `NoSchedule`
   - If you want to schedule a workload one the master (a networking DaemonSet for example), you must tolerate the `node-role.kubernetes.io/master="":NoSchedule` taint
* Supporting only etcd3, with 3.0.14 as the minimum version
   - You can't use etcd2 anymore with kubeadm
* Insecure access to the API Server over `localhost:8080` is disabled by default
   - If you have workloads or other programs that depend on being able to access the master insecurely over localhost, you have to update them to use the secure port (defaulting to 6443)
* The kube-discovery Deployment isn't used anymore when creating a kubeadm cluster and shouldn't be used anywhere else either due to its insecure nature
* The Certificates API has graduated from alpha to beta. This is a backwards-incompatible change since the alpha support is dropped, and therefore kubeadm v1.5 and v1.6 don't work together (for example kubeadm v1.5 can't join a kubeadm v1.6 cluster).

justinsb replied:
This is a great list.  You might want to break them out on the kubeadm page?

BTW I think the etcd v3 we're testing is 3.0.17.

luxas replied:
We have the minimum version set to 3.0.14 still (for external etcd), but use 3.0.17 as the default

luxas commented on release-1.6/release-notes-draft.md:
```diff
@@ -58,12 +58,33 @@
+  service accounts and users are restricted from accessing the API. They can be
+  granted permissions
+  [using the RBAC command-line utilities](https://kubernetes.io/docs/admin/authorization/rbac/#command-line-utilities).
+- kubeadm now uses a new [bootstrap mechanism](https://kubernetes.io/docs/admin/bootstrap-tokens/)
```
Please link to the proposal as well: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/bootstrap-discovery.md, that will be helpful for users in order to understand the flow better.

justinsb commented on release-1.6/release-notes-draft.md:
```diff
@@ -58,12 +58,33 @@
-- Adds a new cloud-controller-manager binary that may be used for testing the new out-of-core cloudprovider flow
-- Introduces an API for the Kubelet to request TLS certificates from the API server.
-- kubeadm is enhanced and improved with a baseline feature set and command line flags that are now marked as beta.
+- kubeadm 1.6 sets up a more secure cluster by default.  It uses [RBAC](https://kubernetes.io/)
```
How about  "_now_ sets up" a more secure cluster by default.  Otherwise it isn't clear what you are comparing it to.

justinsb commented on release-1.6/release-notes-draft.md:
```diff
@@ -58,12 +58,33 @@
 ### New Features
+- kubeadm is a simple tool for creating Kubernetes clusters on existing
+  machines. kubeadm's command line interface is now in beta. This means that
+  using it is considered safe.
```
beta means that you won't change the API without good reason, I think - nothing to do with whether it works or not, for example.  Perhaps drop "This means that" and just go with "Using it is considered safe".

I would still say that a cautionary note is in order here - like "no upgrades, no HA, no cloudprovider support", but OTOH you do have the feedback link in there... On that note, can you please update the link to point to #kubeadm so we aren't overwhelmed in #sig-cluster-lifecycle :-)

lukemarsden commented:
I think I've addressed all review feedback, thanks everyone!

@jbeda @luxas PTAL

pires commented on release-1.6/release-notes-draft.md:
```diff
@@ -58,12 +58,59 @@
+- Added experimental self-hosting support (only available via
+  [config file](https://kubernetes.io/docs/admin/kubeadm/) since it's not
+  reliable yet). This is a step towards making upgrades easier with a planned
+  `kubeadm upgrade` command.
```
Planned where? Can we add links to a proposal or feature issue, if any?

lukemarsden replied:
in our heads :/ I haven't written up a proposal yet.

pires commented on release-1.6/release-notes-draft.md:
```diff
@@ -58,12 +58,59 @@
+  [config file](https://kubernetes.io/docs/admin/kubeadm/) since it's not
+  reliable yet). This is a step towards making upgrades easier with a planned
+  `kubeadm upgrade` command.
+- Added an experimental command for running only one of kubeadm's functions,
```
Can we identify what _kubeadm's functions_ actually are, perhaps link to the phase design document? Also the wording _functions_ doesn't seem right to me, in detriment of _phases_, that is.

luxas replied:
Please link to my phases proposal for this yes. https://github.com/kubernetes/kubeadm/pull/156

lukemarsden replied:
done, and functions -> phases

pires commented on release-1.6/release-notes-draft.md:
```diff
@@ -58,12 +58,59 @@
+  your own, custom deployment. In future versions, this command will mature and
+  be moved to beta quality.
+- kubeadm now uses a new [bootstrap and discovery mechanism](https://kubernetes.io/docs/admin/bootstrap-tokens/)
+  ([proposal](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/bootstrap-discovery.md))
```
kubeadm doesn't use the mechanism proposal, right? Here's how I'd put it:

```
kubeadm now uses a new [bootstrap and discovery mechanism](https://kubernetes.io/docs/admin/bootstrap-tokens/) to form a cluster.
For more details, please take a look at the [design proposal](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/bootstrap-discovery.md).
```

lukemarsden replied:
proposal is in brackets. I want to keep these terse as possible

mikedanese commented on release-1.6/release-notes-draft.md:
```diff
@@ -58,12 +58,59 @@
 ## Cluster Lifecycle
 ### New Features
+- kubeadm is a simple tool for creating Kubernetes clusters on existing
+  machines. kubeadm's command line interface is now in beta.  Using it is
```
Is the entirety of the command line interface in beta? Mention `alpha` subcommand here.

lukemarsden replied:
done

mikedanese commented on release-1.6/release-notes-draft.md:
```diff
@@ -58,12 +58,59 @@
+
 ### Breaking Changes
+- A new label and taint is used for marking the master. The label is
+  `node-role.kubernetes.io/master=""` and the taint has the effect
```
This doesn't say what the taint k/v is.

inercia commented on release-1.6/release-notes-draft.md:
```diff
@@ -58,12 +58,59 @@
+  [RBAC](https://kubernetes.io/) to grant limited privileges to workloads
+  running on the cluster.  By default, service accounts and users are
+  restricted from accessing the API. They can be granted permissions
+  [using the RBAC command-line utilities](https://kubernetes.io/docs/admin/authorization/rbac/#command-line-utilities).
```
https://kubernetes.io/docs/admin/authorization/rbac/#command-line-utilities does not exist

lukemarsden replied:
it will do by the time this is released

idvoretskyi commented:
@lukemarsden please, ensure that the final version of the release notes will be ready today before 1 PM PT today.

mikedanese commented:
@lukemarsden ping.

idvoretskyi commented:
@lukemarsden @mikedanese because of the release delay we can extend the deadline till EOD today.
Otherwise, the notes won't be added.

bgrant0607 commented:
With a quick search, I found the following in the release note draft, which should be listed under breaking changes, with clear explanations of what users need to change:

* kubeadm: kubeadm reset won't drain and remove the current node anymore (#42713, @luxas)
* kubeadm: Rename some flags for beta UI and fixup some logic (#42064, @luxas)
* Flag --use-kubernetes-version for kubeadm init renamed to --kubernetes-version (#41820, @kad)
kubeadm: Remove the --cloud-provider flag for beta init UX (#41710, @luxas)

The following look like they are covered to some degree by this PR:
* Insecure access to the API Server at localhost:8080 will be turned off in v1.6 when using kubeadm (#42066, @luxas)
* Switch to the node-role.kubernetes.io/master label for marking and tainting the master node in kubeadm (#41835, @luxas)
* Add authorization mode to kubeadm (#39846, @andrewrynhard)

However, I suggest you link to RBAC documentation that would better explain what users should expect and what actions they need to take.

cc @jbeda

jbeda commented:
If I don't hear from @lukemarsden in an hour or two I'll pick this up and pull something together by EOD.  @idvoretskyi -- does that work for you?

idvoretskyi commented:
@jbeda yes, please.

lukemarsden commented on release-1.6/release-notes-draft.md:
```diff
@@ -58,12 +58,59 @@
-- Introduces an API for the Kubelet to request TLS certificates from the API server.
-- kubeadm is enhanced and improved with a baseline feature set and command line flags that are now marked as beta.
+- kubeadm 1.6 now sets up a more secure cluster by default.  It uses
+  [RBAC](https://kubernetes.io/) to grant limited privileges to workloads
```
fixed, thanks

calebamiles commented:
@apsinha here's the kubeadm/Cluster Lifecycle release notes PR

lukemarsden commented:
I've pushed a commit which addresses all review comments except:
* I don't know what the taint k/v is, but I agree that it should be mentioned explicitly
* I haven't addressed @bgrant0607's substantive feedback yet

@jbeda I'll take you up on the offer of getting this over the line for me. Please take my WIP fork, squash the commits and make your own PR. Thanks!

@calebamiles please note the above. Thanks!

lukemarsden commented:
Not sure what just happened, I don't think I merged it but GitHub thinks I did. @jbeda sorry could you please make a follow-up PR addressing the above points? Thanks!

Sorry everyone for the tardy response here, I'm in the middle of moving house...

luxas commented:
Copying my earlier comment down from review comments to here so it becomes more visible: https://github.com/kubernetes/features/pull/211#discussion_r106801628

cc @bgrant0607 Here was my take on what we need to have in the relnotes at least

* The kubelet API is now secured, only cluster admins are allowed to access it.
* Insecure access to the API Server over `localhost:8080` is disabled by default
* The control plane components now talk securely to each other. The API Server talks securely to the kubelets in the cluster
* A new label and taint is used for marking the master. The label is `node-role.kubernetes.io/master=""` and the taint has the effect `NoSchedule`
* Added experimental self-hosting support
* Supporting only etcd3, with 3.0.14 as the minimum version
* Added an experimental command for running only one of kubeadm's functions, for instance `kubeadm alpha phase certs selfsign` generates all necessary certificates for a cluster. You can use these building block functions in your own, custom deployment. In future versions, this command will mature and be moved to beta quality.
* Debian packages are built for ppc64le and s390x as well

Some of these are breaking as well:
* A new label and taint is used for marking the master. The label is `node-role.kubernetes.io/master=""` and the taint has the effect `NoSchedule`
   - If you want to schedule a workload one the master (a networking DaemonSet for example), you must tolerate the `node-role.kubernetes.io/master="":NoSchedule` taint
* Supporting only etcd3, with 3.0.14 as the minimum version
   - You can't use etcd2 anymore with kubeadm
* Insecure access to the API Server over `localhost:8080` is disabled by default
   - If you have workloads or other programs that depend on being able to access the master insecurely over localhost, you have to update them to use the secure port (defaulting to 6443)
* The kube-discovery Deployment isn't used anymore when creating a kubeadm cluster and shouldn't be used anywhere else either due to its insecure nature
* The Certificates API has graduated from alpha to beta. This is a backwards-incompatible change since the alpha support is dropped, and therefore kubeadm v1.5 and v1.6 don't work together (for example kubeadm v1.5 can't join a kubeadm v1.6 cluster).

----

@jbeda as @bgrant0607 noted we should also add the flag changes between v1.5 and v1.6

@lukemarsden The taint k/v is the same as the master label k/v: label: `node-role.kubernetes.io/master`, value `""` (simply nothing: ``)

> However, I suggest you link to RBAC documentation that would better explain what users should expect and what actions they need to take.

I think @lukemarsden was about to fix that

lukemarsden commented:
Thanks @luxas, @jbeda will take it from here :)
